### PR TITLE
New Version Refactor

### PR DESCRIPTION
**Pull request by jtenner**
Since the host api is constantly changing, it's a good idea to keep up with the times!

~So far, the changes include `TCPServer`, `connect`, `connect_ip`, and `connect_unsafe` which all create `TCPStream` objects.~

With the lack of multivalue returns, I have included a `Result`-like class which packages the return values of each function into an object on the heap. It uses the new `error` api too, which was very easy to work with in regards to obtaining the result of a host function call.

I believe much more work is needed, and I am thankful for the opportunity to learn and grow as a developer.

TODO:

- [x] implement tcp stream apis
- [x] implement process apis
- [x] implement message apis (still need to learn how to use them)
- [x] refactor host bindings into their own module
- [x] clean up source code formatting, and switch to snake case for everything
- [x] documentation! The readme doesn't get enough love

### CODE REVIEW

File: assembly/net/index.ts [hunk 1]
```diff
@@ -14,8 +17,8 @@ import { iovec } from "bindings/wasi";
 declare function lunatic_resolve(
   name_ptr: usize /* *const u8 */,
   name_len: usize,
-  resolver_id: usize /* *mut u32 */
-): TCPErrorCode;
+  resolver_id: usize /* *mut u64 */
```
Comment (withtypes) [inline assembly/net/index.ts]: Doc string needs to updated, it refers to `u32` instead of `u64`.

File: assembly/net/index.ts [hunk 2]
```diff
@@ -14,8 +17,8 @@ import { iovec } from "bindings/wasi";
 declare function lunatic_resolve(
```
Comment (withtypes) [inline assembly/net/index.ts]: @bkolobara is the Wasm module name now `lunatic::networking` for networking functions? It means in AS `@external("lunatic", "resolve")` should be `@external("lunatic::networking", "resolve")`?

File: assembly/net/index.ts [hunk 3]
```diff
@@ -27,208 +30,156 @@ const enum ResolveNextResult {
...
-  /** Some other kind of error has occurred. */
-  Other = 20,
-}
-
 /**
  * Write the next IP address into memory if it exists.
  *
- * @param {u32} resolver_id - The iterator id.
+ * @param {u64} resolver_id - The iterator id.
  * @param {usize} addr - A pointer to 16 allocated bytes to write the resolved IP address.
```
Comment (withtypes) [inline assembly/net/index.ts]: It think https://github.com/lunatic-solutions/rust-lib/blob/main/src/host_api.rs#L30 has `addr_type` first and then `addr`.

File: assembly/net/index.ts [hunk 4]
```diff
@@ -27,208 +30,156 @@ const enum ResolveNextResult {
...
 
 
 /**
  * Bind a TCPServer to an IP Address and port.
  *
- * @param {usize} addr_ptr - A pointer to the address bytes.
  * @param {usize} addr_len - The length of the IP address, 4 or 16.
+ * @param {usize} addr_ptr - A pointer to the address bytes.
  * @param {u16} port - The port.
  * @param {usize} listener_id - A pointer to a u32 that will be the TCPServer
```
Comment (withtypes) [inline assembly/net/index.ts]: Pointer to `u64`

File: assembly/net/index.ts [hunk 5]
```diff
@@ -237,337 +188,407 @@ declare function tcp_stream_deserialize(tcp_stream: u32): u32;
 // @ts-ignore: valid decorator
 @external("lunatic", "tcp_bind")
 declare function tcp_bind(
-  addr_ptr: usize,// *const u8,
   addr_len: usize,
+  addr_ptr: usize,// *const u8,
   port: u16,
+  flow_info: u32,
+  scope_id: u32,
   listener_id: usize, //*mut u32,
```
Comment (withtypes) [inline assembly/net/index.ts]: `*mut u64`

File: assembly/net/index.ts [hunk 6]
```diff
@@ -237,337 +188,407 @@ declare function tcp_stream_deserialize(tcp_stream: u32): u32;
...
+ * @param {usize} listener_id - A pointer to a u32 where the listener_id will be written if
+ * the connection was successful, or the error id if it failed.
+ * @returns {TCPErrorCode} The result of initiating a TCP connection.
  */
 // @ts-ignore: valid decorator
-@external("lunatic", "tcp_listener_serialize")
-declare function tcp_listener_serialize(tcp_listener: u32): u32;
+@external("lunatic", "tcp_connect")
+declare function tcp_connect(
+  addr_ptr: usize, // *const u8,
```
Comment (withtypes) [inline assembly/net/index.ts]: I think the order of `addr_ptr` and `addr_len` was reversed. https://github.com/lunatic-solutions/rust-lib/blob/main/src/host_api.rs#L46

File: assembly/net/index.ts [hunk 7]
```diff
@@ -27,208 +30,156 @@ const enum ResolveNextResult {
...
-  tcp_stream: u32,
-  data: usize, // *const c_void,
-  data_len: usize,
-  nwritten: usize // *mut usize,
-): TCPErrorCode;
-
-/** Flush a tcp stream. */
-// @ts-ignore: valid decorator
-@external("lunatic", "tcp_flush")
-declare function tcp_flush(tcp_stream: u32): TCPErrorCode;
```
Comment (withtypes) [inline assembly/net/index.ts:197]: There is `tcp_flush` missing?

**Review by withtypes — COMMENTED** [review]
I compared the interfaces. I didn't try to run it or anything.

**Comment by jtenner** [comment]
This is all still a work in progress. I should definitely add tcp_flush for sure!

Also, any result that comes back from lunatic needs to be a u64 because error ids are 64 bits.

File: assembly/net/index.ts (hunk 3 above)
Comment (jtenner) [reply assembly/net/index.ts]: Good catch!

File: assembly/net/index.ts (hunk 2 above)
Comment (bkolobara) [reply assembly/net/index.ts]: Correct!

**Comment by bkolobara** [comment]
Not only errors ids, but all IDs are now 64bit.

File: assembly/net/index.ts (hunk 6 above)
Comment (bkolobara) [reply assembly/net/index.ts]: There is no more `addr_len` (4 or 16), only `addr_type` (4 or 6, refering to the IPv4 or IPv6 (16 bytes) formats)

**Comment by bkolobara** [comment]
@jtenner I had a call today with @withtypes, where we discussed some future developments. Some of the APIs (especially around message sending) may change again. I would also like to have a global registry of resources with `UUID` IDs so you can easily send them between processes without the "transform to local ID" dance we currently do. This should make the guest implementation around serialisation way easier.

Just giving you a heads up if you want to wait a bit before everything is ironed out to complete the re-write.

P.S. We are also looking into expressing some of the APIs through an "everything is a process" approach. For example binding to a socket would mean that you get a `process` back and you can send messages to it. These messages would automatically be turned into network I/O. This is still in the drafting phase.

**Comment by jtenner** [comment]
Given all of these changes, and the future changes coming down the pipeline, I think it would be wise for me to wait and re-create all these apis when everything settles down. Thank you for the heads up! Can you message me when everything is more settled down? :)

**Comment by gdamjan** [comment]
could the low-level bindings be automatically generated?

**Comment by jtenner** [comment]
I don't think that could be a problem. I don't entirely know what needs to be done, but I might be making my return soon to get this software working properly :). It might be time to get working.

File: assembly/net/index.ts (hunk 6 above)
Comment (jtenner) [reply assembly/net/index.ts:230]: Hey guys. Is this an `u32` then?

**Comment by jtenner** [comment]
Please feel free to take my work and run with it. I appreciate everyone's time and effort.

**Comment by jtenner** [comment]
Need to start writing tests and validating how things work. Anyone can contribute tests

**Comment by jtenner** [comment]
Okay! Let's follow up with some comments about what has happened so far:

1. After months of work, as-lunatic is finally up to snuff to the point where testing should be done. Given how everything went, I decided to go against my principals and do testing last which is lazy. I'm sorry it ended like this.
2. Most of the APIs appear to be implemented along with a very custom and currently unstable custom sized static hashmap for finalization.
3. The default count for finalization records is about 1000 records max in total. The program essentially has a c-array of finalization records in static memory below the heap base that help prevent the need for free and malloc while finalization is occuring.
4. ~I believe I can resize the heap allocation for the map, but that will come later.~
5. The code formatting is less than what I like. Not sure if I should follow JavaScript standard casing or snake case where appropriate. Either way, it's inconsistent.
6. Docs are needed. Badly.

**Comment by jtenner** [comment]
I decided that using heap.allocate during hashmap sets was safe because allocating a reference at construction time is safe. As a result, the hashtable can be resized. This caused a few distinct problems with the hashmap algorithm, and I had to implement a `free` property on the entry.

It's calling allocate or free mid garbage collection that causes an issue. So this is now optimized a little better.

**Comment by jtenner** [comment]
After finally doing more code comments and upgrading CI to install latest lunatic, node, AssemblyScript, and run the test script, CI passes. This is probably not robust enough testing, and more work is necessary, but this is a great stopping point for the night.

**Comment by jtenner** [comment]
Looks like I should handle the new process version function calls too, but that can be done on a later day. Today I'm going to make a telnet echo server using as-telnet to test the TCP connection logic.

**Comment by jtenner** [comment]
Leftover problems are all `INFO` or `WARNING`s.

```
WARNING AS201: Conversion from type 'usize' to 'i32' will require an explicit cast when switching between 32/64-bit.

       sum += changetype<iovec>(vec + i * offsetof<iovec>()).buf_len;
              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 in ~lib/as-lunatic/util/index.ts(120,14)

INFO AS210: Expression is never 'null'.

       assert(false, message ? message! : this.errStr!);
                               ~~~~~~~
 in ~lib/as-lunatic/error/index.ts(62,31)

INFO AS210: Expression is never 'null'.

       assert(false, message ? message! : this.errStr!);
                               ~~~~~~~
 in ~lib/as-lunatic/error/index.ts(62,31)


> as-lunatic-telnet-echo@1.0.0 asbuild:optimized
> asc assembly/index.ts --target release

WARNING AS201: Conversion from type 'usize' to 'i32' will require an explicit cast when switching between 32/64-bit.

       sum += changetype<iovec>(vec + i * offsetof<iovec>()).buf_len;
              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 in ~lib/as-lunatic/util/index.ts(120,14)

INFO AS210: Expression is never 'null'.

       assert(false, message ? message! : this.errStr!);
                               ~~~~~~~
 in ~lib/as-lunatic/error/index.ts(62,31)

INFO AS210: Expression is never 'null'.

       assert(false, message ? message! : this.errStr!);
                               ~~~~~~~
 in ~lib/as-lunatic/error/index.ts(62,31)
 ```

Looks like I could use to clean these up a little bit.

**Comment by jtenner** [comment]
This is now ready for review and alpha release.

**Comment by jtenner** [comment]
A lot of the necessary features are here. I am currently implementing tasks, but there is one issue left that I want to address before all of this becomes real.

```ts
// this should be a compile time error
let a = new Process<T>();
```

I need to do some refactoring to make sure that end users only use the static `Process` and `TCPServer` methods to construct sockets, servers, and create processes.

**Pull request merged**